cmd/main.go

```diff
@@ -0,0 +1,196 @@
+	return func(ctx *cli.Context) error {
+		var args setting.ArgWorkPathAndCustomConf
+		ctxLineage := ctx.Lineage()
+		for i := len(ctxLineage) - 1; i >= 0; i-- {
+			curCtx := ctxLineage[i]
+			if curCtx.IsSet("work-path") && args.WorkPath == "" {
+				args.WorkPath = curCtx.String("work-path")
+			}
+			if curCtx.IsSet("custom-path") && args.CustomPath == "" {
+				args.CustomPath = curCtx.String("custom-path")
+			}
+			if curCtx.IsSet("config") && args.CustomConf == "" {
+				args.CustomConf = curCtx.String("config")
+			}
+		}
+		setting.InitWorkPathAndCommonConfig(os.Getenv, args)
+		if ctx.Bool("help") || action == nil {
+			// the default behavior of "urfave/cli": "nil action" means "show help"
+			return cmdHelp().Action(ctx)
+		}
+		return action(ctx)
+	}
+}
+
+func reflectGet(v any, fieldName string) any {
+	e := reflect.ValueOf(v).Elem()
+	return e.FieldByName(fieldName).Interface()
+}
+
+// https://cli.urfave.org/migrate-v1-to-v2/#flag-aliases-are-done-differently
+// Sadly v2 doesn't warn you if a comma is in the name. (https://github.com/urfave/cli/issues/1103)
+func checkCommandFlags(c any) bool {
+	var cmds []*cli.Command
+	if app, ok := c.(*cli.App); ok {
+		cmds = app.Commands
+	} else {
+		cmds = c.(*cli.Command).Subcommands
+	}
+	ok := true
+	for _, cmd := range cmds {
+		for _, flag := range cmd.Flags {
+			flagName := reflectGet(flag, "Name").(string)
+			if strings.Contains(flagName, ",") {
+				ok = false
+				log.Error("cli.Flag can't have comma in its Name: %q, use Aliases instead", flagName)
+			}
+		}
+		if !checkCommandFlags(cmd) {
+			ok = false
+		}
+	}
+	return ok
+}
+
+func NewMainApp() *cli.App {
+	app := cli.NewApp()
+	app.EnableBashCompletion = true
+
+	// these sub-commands need to use config file
+	subCmdWithConfig := []*cli.Command{
+		CmdWeb,
+		CmdServ,
+		CmdHook,
+		CmdDump,
+		CmdAdmin,
+		CmdMigrate,
+		CmdKeys,
+		CmdConvert,
+		CmdDoctor,
+		CmdManager,
+		CmdEmbedded,
+		CmdMigrateStorage,
+		CmdDumpRepository,
+		CmdRestoreRepository,
+		CmdActions,
+		cmdHelp(), // the "help" sub-command was used to show the more information for "work path" and "custom config"
+	}
+
+	// these sub-commands do not need the config file, and they do not depend on any path or environment variable.
+	subCmdStandalone := []*cli.Command{
+		CmdCert,
+		CmdGenerate,
+		CmdDocs,
+	}
+
+	app.DefaultCommand = CmdWeb.Name
+
+	globalFlags := appGlobalFlags()
+	app.Flags = append(app.Flags, globalFlags...)
+	app.HideHelp = true // use our own help action to show helps (with more information like default config)
+	app.Before = PrepareConsoleLoggerLevel(log.INFO)
+	for i := range subCmdWithConfig {
+		prepareSubcommandWithConfig(subCmdWithConfig[i], globalFlags)
+	}
+	app.Commands = append(app.Commands, subCmdWithConfig...)
+	app.Commands = append(app.Commands, subCmdStandalone...)
+
+	if !checkCommandFlags(app) {
+		panic("some flags are incorrect") // this is a runtime check to help developers
```

Comment:
Should we perhaps print exactly which flags are incorrect?

Reply:
There are details in `checkCommandFlags`

> log.Error("cli.Flag can't have comma in its Name: %q, use Aliases instead", flagName)


This "panic" only panics.